lib/pub_sub/breaker.rb

```diff
@@ -0,0 +1,55 @@
+module PubSub
+
+  class Breaker
+
+    THREAD_LOCAL_IDENTIFIER = :pub_sub_region
+
+    class << self
+
+      def run(&block)
+        current_breaker.run(&block)
+      rescue CB2::BreakerOpen
+        Breaker.use_next_breaker
+        sleep 1
+        retry
+      end
+
+      def current_breaker
+        all_breakers[current_breaker_idx]
+      end
+
+      def current_region
+        all_regions[current_breaker_idx]
+      end
+
+      def use_next_breaker
+        Thread.current[THREAD_LOCAL_IDENTIFIER] += 1
+        Thread.current[THREAD_LOCAL_IDENTIFIER] = Thread.current[THREAD_LOCAL_IDENTIFIER] % all_breakers.count
+      end
+
+      private
+
+      def current_breaker_idx
+        Thread.current[THREAD_LOCAL_IDENTIFIER] ||= 0
+      end
+
+      def all_breakers
+        @breakers ||= all_regions.map do |region|
```

Comment:
Good catch. This won't be a problem in this case because it's memoized for speed rather than correctness, but you're 100% right, the breakers could be initialized twice.